auth/auth_test.go

```diff
@@ -506,3 +506,84 @@ func TestNew2LOTokenProvider_Validate(t *testing.T) {
 		})
 	}
 }
+
+type countingTestProvider struct {
+	count int
+}
+
+func (tp *countingTestProvider) Token(ctx context.Context) (*Token, error) {
+	tok := &Token{
+		Value: fmt.Sprint(tp.count),
+		// Set expiry to q1 second from now.
+		Expiry: time.Now().Add(time.Second),
+	}
+	tp.count++
+	return tok, nil
+}
+func TestComputeTokenProvider_NonBlockingRefresh(t *testing.T) {
+	// Freeze now for consistent results.
+	now := time.Now()
+	timeNow = func() time.Time { return now }
+	defer func() { timeNow = time.Now }()
+	tp := NewCachedTokenProvider(&countingTestProvider{count: 1}, &CachedTokenProviderOptions{
+		// EarlyTokenRefresh ensures that token with expiry 1 second from now is already stale.
+		ExpireEarly: 2 * time.Second,
+	})
+	freshToken, err := tp.Token(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+	if want := "1"; freshToken.Value != want {
```

Comment:
nit: can this test or another test be expanded to show the token hitting the different states? Since this is an internal test we should be able to safely access those internal bits. If done as a separate test you could keep changing timeNow() to make sure we hit all the states.

Reply:
Done, thanks!